### DISCUSSION TITLE
Lead & lag broken?

### DISCUSSION THREAD

**Discussion by Quirq-UK** [hydrogen-music/hydrogen#1662] [Hydrogen on Linux]
Hi

Lead and lag appears to be broken in the version that I'm using, well, the lag part at least. 

It didn't sound like it was doing anything so I recorded some extreme lead/lag variations in Ardour and zoomed right in: the hits with full lead were ahead of the beat as expected, but the hits with full lag appeared dead on the beat rather than behind. Is this a known issue?

I'm using 1.1.1-'be336a1d' on UbuntuStudio 20.04.5 LTS.

Also, I remember from ages ago that lead and lag didn't appear in MIDI files exported from Hydrogen and this still appears to be the case. Is this ever likely to change i.e. lead and lag information will be included in MIDI exports?

**Comment by theGreatWhiteShark** [comment]
Hey @Quirq-UK ,

Thanks for the notice!

> Lead and lag appears to be broken in the version that I'm using, well, the lag part at least.

This was already reported recently (#1655) and will be fixed in the upcoming release.

> Also, I remember from ages ago that lead and lag didn't appear in MIDI files exported from Hydrogen and this still appears to be the case. Is this ever likely to change i.e. lead and lag information will be included in MIDI exports?

Yeah. Audio and MIDI processing are two entirely different things within Hydrogen. I never looked into the MIDI part yet. But maybe this can be introduced while fixing the former. Not sure yet. I'll keep you posted.

**Reply by Quirq-UK** [reply]
> This was already reported recently (https://github.com/hydrogen-music/hydrogen/issues/1655) and will be fixed in the upcoming release.

Ah, okay, I'd seen that bug but hadn't realised that it also applied to the lead and lag! That's good to know there's a fix on the way, thanks.

> Yeah. Audio and MIDI processing are two entirely different things within Hydrogen.

It never occurred to me that the lead/lag was applied at the audio level rather than to the MIDI. It not affecting the MIDI is a problem for me because of my workflow: I use Hydrogen purely for programming and have the drum samples loaded in LinuxSampler on a track in Ardour. I hook Hydrogen up to Ardour whilst I'm writing drum parts, then export the finished MIDI, import into Ardour rendering as audio (I use a lot of bpm ramps and it can get a little glitchy doing it any other way). It'd be great to have the lead and lag imprinted on the MIDI.

By the way, I'd not used Hydrogen in a few years and came back to it recently. I must say all the cut/paste features are really good and _very_ useful, thanks to all involved!

**Reply by cme** [reply]
> > Yeah. Audio and MIDI processing are two entirely different things within Hydrogen.
> 
> It never occurred to me that the lead/lag was applied at the audio level rather than to the MIDI. It not affecting the MIDI is a problem for me because of my workflow: I use Hydrogen purely for programming and have the drum samples loaded in LinuxSampler on a track in Ardour. I hook Hydrogen up to Ardour whilst I'm writing drum parts, then export the finished MIDI, import into Ardour rendering as audio (I use a lot of bpm ramps and it can get a little glitchy doing it any other way). It'd be great to have the lead and lag imprinted on the MIDI.

The lead/lag (and swing and humanisation) are all at the sample level, and so potentially have much finer-grained timing than exporting to a MIDI file would have. I guess we'd have to quantise the time and round to the nearest tick. That might be a problem because Hydrogen's TPQN is lower than a lot of DAWs use I think so the changes might be more noticable that folk would like. Changing it in Hydrogen is probably not trivial :/

> By the way, I'd not used Hydrogen in a few years and came back to it recently. I must say all the cut/paste features are really good and _very_ useful, thanks to all involved!

I am very glad to hear this. Thanks :)

**Reply by Quirq-UK** [reply]
Thinking about it, it seems it must be affecting the MIDI in some way because I can record lead (but not lag) when I record Hydrogen's MIDI output in Ardour, but it just doesn't affect MIDI when Hydrogen renders to a file internally.

**Reply by theGreatWhiteShark** [reply]
> Thinking about it, it seems it must be affecting the MIDI in some way because I can record lead (but not lag) when I record Hydrogen's MIDI output in Ardour, but it just doesn't affect MIDI when Hydrogen renders to a file internally.

Yes. MIDI notes sent during playback are handled by the audio engine/sampler and feature the same humanization as those rendered to audio. MIDI export, however, is basically a function that converts the song file into a MIDI file.

> That might be a problem because Hydrogen's TPQN is lower than a lot of DAWs use I think so the changes might be more noticable that folk would like. Changing it in Hydrogen is probably not trivial :/

Is it? In the rewritten audio engine for the upcoming 1.2. version ticks are handled as doubles. It is true that our tick sizes (number of frames that constitute a tick) are larger than in most DAWs. But with smaller patches in `AudioEngine::updateNoteQueue` and in places using `Note::__position` we should be able to support note placement with double precision (and true tuplets etc) as well. I may mistaking but isn't the grid resolution in the Pattern Editor the only thing left that requires our TPQN and keeps us from using much more fine-grained precision?

At least right now I don't see why we should be restricted to our tick sizes in the MIDI export instead of the lowest resolution offered by the MIDI format.

**Comment by theGreatWhiteShark** [comment]
Hey @Quirq-UK ,

Lead & Lag should work again in #1669.

But I decided against introducing humanization into MIDI export at this point. However, it is on the TODO list.
